File: receiver_android/app/src/test/java/com/iflytek/lanmediacast/receiver/protocol/ProtocolFixtureTest.kt
package com.iflytek.lanmediacast.receiver.protocol

import java.io.File
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.jsonArray
import org.junit.Assert.assertEquals
import org.junit.Assert.assertThrows
import org.junit.Test

class ProtocolFixtureTest {
    private val fixtures = File("../../protocol/fixtures/v1")

    @Test
    fun `valid discovery fixtures decode`() {
        val query = ProtocolCodec.decodeDiscoveryQuery(read("valid/discover_query.json").toByteArray())
        assertEquals("Teacher-PC", query.senderName)
    }

    @Test
    fun `valid envelope fixtures decode`() {
        assertEquals("player.play", ProtocolCodec.decodeEnvelope(read("valid/command_envelope.json")).type)
        assertEquals("protocol.error", ProtocolCodec.decodeEnvelope(read("valid/protocol_error.json")).type)
        assertEquals(
            "internal_error",
            ProtocolCodec.decodeEnvelope(read("valid/protocol_internal_error.json"))
                .payload["reason"]?.jsonPrimitive?.content,
        )
        assertEquals("response", ProtocolCodec.decodeEnvelope(read("valid/response_envelope.json")).type)
        val ready = ProtocolCodec.decodeEnvelope(read("valid/session_ready.json"))
        assertEquals("Projector-01", ready.payload["deviceName"]?.jsonPrimitive?.content)
        assertEquals("12", ready.payload["playlistRevision"]?.jsonPrimitive?.content)
        assertEquals(
            "pairing_expired",
            ProtocolCodec.decodeEnvelope(read("valid/pairing_expired_response.json"))
                .payload["error"]?.jsonObject?.get("code")?.jsonPrimitive?.content,
        )
        assertEquals("photo.batch.ready", ProtocolCodec.decodeEnvelope(read("valid/photo_batch_ready.json")).type)
        assertEquals("photo.item.ready", ProtocolCodec.decodeEnvelope(read("valid/photo_item_ready.json")).type)
        validatePhotoResume(ProtocolCodec.decodeEnvelope(read("valid/photo_batch_resume_state.json")).payload)
        assertEquals(
            "unknown_transfer",
            ProtocolCodec.decodeEnvelope(read("valid/photo_protocol_unknown_transfer.json"))
                .payload["reason"]?.jsonPrimitive?.content,
        )
        assertEquals(
            "null",
            ProtocolCodec.decodeEnvelope(read("valid/photo_protocol_malformed_header.json"))
                .payload["transferId"]?.jsonPrimitive?.content,
        )
    }

    @Test
    fun `invalid fixtures are rejected`() {
        assertThrows(ProtocolException::class.java) {
            ProtocolCodec.decodeDiscoveryQuery(read("invalid/discover_wrong_version.json").toByteArray())
        }
        listOf("command_missing_sequence.json", "invalid_message_type.json", "protocol_error_with_reply.json")
            .forEach { name ->
            assertThrows(ProtocolException::class.java) {
                ProtocolCodec.decodeEnvelope(read("invalid/$name"))
            }
        }
        assertThrows(IllegalArgumentException::class.java) {
            validatePhotoResume(
                ProtocolCodec.decodeEnvelope(read("invalid/photo_resume_missing_nullable.json")).payload,
            )
        }
    }

    @Test
    fun `remote HTTP header fixtures enforce the shared allowlist`() {
        val valid = ProtocolCodec.json.parseToJsonElement(read("valid/remote_http_headers.json")).jsonObject
        assertEquals(
            listOf("User-Agent", "Referer", "Origin", "Accept", "Accept-Language"),
            RemoteHttpHeaders.fromSource(
                kotlinx.serialization.json.buildJsonObject { put("httpHeaders", valid) },
            ).keys.toList(),
        )

        val invalid = ProtocolCodec.json.parseToJsonElement(read("invalid/remote_http_headers.json")).jsonObject
        invalid.getValue("cases").jsonArray.forEach { headers ->
            assertThrows(IllegalArgumentException::class.java) {
                RemoteHttpHeaders.fromSource(
                    kotlinx.serialization.json.buildJsonObject { put("httpHeaders", headers) },
                )
            }
        }
    }

    @Test
    fun `split remote media source fixtures are synchronized`() {
        val valid = ProtocolCodec.json.parseToJsonElement(
            read("valid/split_remote_media_source.json"),
        ).jsonObject
        val normalized = RemoteMediaSourceValidator.validate(valid)
        assertEquals("web:example:primary", normalized.primaryTrack.source["cacheKey"]?.jsonPrimitive?.content)
        assertEquals("web:example:audio", normalized.audioTrack?.source?.get("cacheKey")?.jsonPrimitive?.content)

        val invalid = ProtocolCodec.json.parseToJsonElement(
            read("invalid/split_remote_media_source.json"),
        ).jsonObject
        invalid.getValue("cases").jsonArray.forEach { source ->
            assertThrows(IllegalArgumentException::class.java) {
                RemoteMediaSourceValidator.validate(source.jsonObject)
            }
        }
    }

    private fun validatePhotoResume(payload: kotlinx.serialization.json.JsonObject) {
        val expected = setOf("awaitingMeta", "ready", "partial", "complete", "removed")
        val seen = payload.getValue("items").jsonArray.map { element ->
            val item = element.jsonObject
            require(item.containsKey("transferId") && item.containsKey("nextChunkIndex")) {
                "nullable fields must be present"
            }
            item.getValue("status").jsonPrimitive.content.also { require(it in expected) }
        }.toSet()
        require(seen == expected) { "status fixture coverage incomplete" }
    }

    private fun read(relativePath: String): String {
        val file = File(fixtures, relativePath)
        check(file.isFile) { "Missing protocol fixture: ${file.absolutePath}" }
        return file.readText(Charsets.UTF_8)
    }
}

File: sender_flutter/test/protocol_fixture_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:lan_media_cast_sender/protocol/protocol.dart';
import 'package:lan_media_cast_sender/protocol/remote_http_headers.dart';
import 'package:lan_media_cast_sender/protocol/remote_media_source.dart';

void main() {
  final Directory fixtures = Directory('../protocol/fixtures/v1');

  String fixture(String relativePath) {
    final File file = File('${fixtures.path}/$relativePath');
    expect(file.existsSync(), isTrue, reason: 'Missing fixture: ${file.path}');
    return file.readAsStringSync();
  }

  test('valid discovery fixtures decode', () {
    final DiscoveryQuery query = DiscoveryQuery.decode(
      File('${fixtures.path}/valid/discover_query.json').readAsBytesSync(),
    );
    expect(query.senderName, 'Teacher-PC');

    final DiscoveryResponse response = DiscoveryResponse.decode(
      File('${fixtures.path}/valid/discover_response.json').readAsBytesSync(),
    );
    expect(response.wssPort, 39881);
  });

  test('valid envelope fixtures decode', () {
    expect(
      ProtocolEnvelope.decode(fixture('valid/command_envelope.json')).type,
      'player.play',
    );
    expect(
      ProtocolEnvelope.decode(fixture('valid/protocol_error.json')).type,
      'protocol.error',
    );
    expect(
      ProtocolEnvelope.decode(
        fixture('valid/protocol_internal_error.json'),
      ).payload['reason'],
      'internal_error',
    );
    expect(
      ProtocolEnvelope.decode(fixture('valid/response_envelope.json')).type,
      'response',
    );
    final ProtocolEnvelope ready = ProtocolEnvelope.decode(
      fixture('valid/session_ready.json'),
    );
    expect(ready.payload['deviceName'], 'Projector-01');
    expect(ready.payload['playlistRevision'], 12);
    expect(
      ProtocolEnvelope.decode(
        fixture('valid/pairing_expired_response.json'),
      ).payload['error'],
      containsPair('code', 'pairing_expired'),
    );
    expect(
      ProtocolEnvelope.decode(fixture('valid/photo_batch_ready.json')).type,
      'photo.batch.ready',
    );
    expect(
      ProtocolEnvelope.decode(fixture('valid/photo_item_ready.json')).type,
      'photo.item.ready',
    );
    expect(
      ProtocolEnvelope.decode(fixture('valid/receiver_logs_get.json')).type,
      'diagnostics.logs.get',
    );
    final ProtocolEnvelope resume = ProtocolEnvelope.decode(
      fixture('valid/photo_batch_resume_state.json'),
    );
    _validatePhotoResumePayload(resume.payload);
    expect(
      ProtocolEnvelope.decode(
        fixture('valid/photo_protocol_unknown_transfer.json'),
      ).payload['reason'],
      'unknown_transfer',
    );
    expect(
      ProtocolEnvelope.decode(
        fixture('valid/photo_protocol_malformed_header.json'),
      ).payload['transferId'],
      isNull,
    );
  });

  test('invalid fixtures are rejected', () {
    expect(
      () => DiscoveryQuery.decode(
        File(
          '${fixtures.path}/invalid/discover_wrong_version.json',
        ).readAsBytesSync(),
      ),
      throwsA(isA<ProtocolException>()),
    );
    for (final String name in [
      'command_missing_sequence.json',
      'invalid_message_type.json',
      'protocol_error_with_reply.json',
    ]) {
      expect(
        () => ProtocolEnvelope.decode(fixture('invalid/$name')),
        throwsA(isA<ProtocolException>()),
      );
    }
    final ProtocolEnvelope invalidResume = ProtocolEnvelope.decode(
      fixture('invalid/photo_resume_missing_nullable.json'),
    );
    expect(
      () => _validatePhotoResumePayload(invalidResume.payload),
      throwsFormatException,
    );
  });

  test('remote HTTP header fixtures enforce the shared allowlist', () {
    final Map<String, dynamic> valid =
        jsonDecode(fixture('valid/remote_http_headers.json'))
            as Map<String, dynamic>;
    expect(validateRemoteHttpHeaders(valid).keys, <String>[
      'User-Agent',
      'Referer',
      'Origin',
      'Accept',
      'Accept-Language',
    ]);

    final Map<String, dynamic> invalid =
        jsonDecode(fixture('invalid/remote_http_headers.json'))
            as Map<String, dynamic>;
    for (final Object? rawCase in invalid['cases']! as List<dynamic>) {
      expect(() => validateRemoteHttpHeaders(rawCase), throwsFormatException);
    }
  });

  test('split remote media source fixtures are synchronized', () {
    final Map<String, dynamic> valid =
        jsonDecode(fixture('valid/split_remote_media_source.json'))
            as Map<String, dynamic>;
    final Map<String, Object> normalized = validateRemoteMediaSource(valid);
    expect(normalized['cacheKey'], 'web:example:primary');
    expect(
      normalized['audioTrack'],
      isA<Map<String, Object>>().having(
        (Map<String, Object> track) => track['cacheKey'],
        'cacheKey',
        'web:example:audio',
      ),
    );

    final Map<String, dynamic> invalid =
        jsonDecode(fixture('invalid/split_remote_media_source.json'))
            as Map<String, dynamic>;
    for (final Object? rawCase in invalid['cases']! as List<dynamic>) {
      expect(() => validateRemoteMediaSource(rawCase), throwsFormatException);
    }
  });
}

void _validatePhotoResumePayload(Map<String, dynamic> payload) {
  final Object? rawItems = payload['items'];
  if (rawItems is! List<dynamic>) throw const FormatException('items missing');
  const Set<String> statuses = <String>{
    'awaitingMeta',
    'ready',
    'partial',
    'complete',
    'removed',
  };
  final Set<String> seen = <String>{};
  for (final Object? rawItem in rawItems) {
    if (rawItem is! Map<String, dynamic> ||
        !rawItem.containsKey('transferId') ||
        !rawItem.containsKey('nextChunkIndex')) {
      throw const FormatException('nullable fields must be present');
    }
    final Object? status = rawItem['status'];
    if (status is! String || !statuses.contains(status)) {
      throw const FormatException('invalid status');
    }
    seen.add(status);
  }
  if (seen.length != statuses.length) {
    throw const FormatException('status fixture coverage incomplete');
  }
}
